server/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java

```diff
@@ -39,71 +45,150 @@ public class DfsPhase {
+                if (context.isCancelled()) {
+                    throw new TaskCancelledException("cancelled");
+                }
+                CollectionStatistics cs = super.collectionStatistics(field);
+                if (cs != null) {
+                    fieldStatistics.put(field, cs);
                 }
+                return cs;
             }
+        };
 
-            Term[] terms = stats.keySet().toArray(new Term[0]);
-            TermStatistics[] termStatistics = new TermStatistics[terms.length];
-            for (int i = 0; i < terms.length; i++) {
-                termStatistics[i] = stats.get(terms[i]);
+        searcher.createWeight(context.rewrittenQuery(), ScoreMode.COMPLETE, 1);
+        for (RescoreContext rescoreContext : context.rescore()) {
+            for (Query query : rescoreContext.getQueries()) {
+                searcher.createWeight(context.searcher().rewrite(query), ScoreMode.COMPLETE, 1);
             }
+        }
 
-            context.dfsResult()
-                .termsStatistics(terms, termStatistics)
-                .fieldStatistics(fieldStatistics)
-                .maxDoc(context.searcher().getIndexReader().maxDoc());
+        Term[] terms = stats.keySet().toArray(new Term[0]);
+        TermStatistics[] termStatistics = new TermStatistics[terms.length];
+        for (int i = 0; i < terms.length; i++) {
+            termStatistics[i] = stats.get(terms[i]);
+        }
+
+        context.dfsResult()
+            .termsStatistics(terms, termStatistics)
+            .fieldStatistics(fieldStatistics)
+            .maxDoc(context.searcher().getIndexReader().maxDoc());
+    }
+
+    private void executeKnnVectorQuery(SearchContext context) throws IOException {
+        SearchSourceBuilder source = context.request().source();
+        if (source == null || source.knnSearch() == null) {
+            return;
+        }
 
-            // If kNN search is requested, perform kNN query and gather top docs
-            SearchSourceBuilder source = context.request().source();
-            if (source != null && source.knnSearch() != null) {
-                SearchExecutionContext searchExecutionContext = context.getSearchExecutionContext();
-                KnnSearchBuilder knnSearch = source.knnSearch();
+        boolean profile = context.getProfilers() != null;
+        long knnStartTime = profile ? System.nanoTime() : 0L;
+        final Timer rewriteTimer = profile ? new Timer() : null;
+        final Timer createWeightTimer = profile ? new Timer() : null;
 
-                KnnVectorQueryBuilder knnVectorQueryBuilder = knnSearch.toQueryBuilder();
-                if (context.request().getAliasFilter().getQueryBuilder() != null) {
-                    knnVectorQueryBuilder.addFilterQuery(context.request().getAliasFilter().getQueryBuilder());
+        IndexSearcher searcher = new IndexSearcher(context.searcher().getIndexReader()) {
```

Comment:
should this be `profile ? <your subclass> : context.searcher()`?

Reply:
It'd be nice to add a comment here explaining why we can't just use the context searcher in both cases. I assume it's because we don't want to use the `QueryProfiler` that it uses?